sdk/appconfiguration/azure-appconfiguration-provider/azure/appconfiguration/provider/_azureappconfigurationprovider.py

```diff
@@ -352,3 +472,47 @@ def __exit__(self, *args) -> None:
         self._client.__exit__(*args)
         for client in self._secret_clients.values():
             client.__exit__()
+
+    class _ConfigurationRefresh:
+        def __init__(self, **kwargs):
+            self.refresh_options: Optional[AzureAppConfigurationRefreshOptions] = kwargs.pop(
+                "refresh_options", AzureAppConfigurationRefreshOptions()
+            )
+            if self.refresh_options is not None:
+                self.next_refresh_time = datetime.now() + timedelta(seconds=self.refresh_options.refresh_interval)
+            self.attempts = 1
+            self.min_backoff: Optional[int] = kwargs.pop("min_backoff", 30)
+            self.max_backoff: Optional[int] = kwargs.pop("min_backoff", 600)
+
+        def updated_configurations(self):
+            # pylint:disable=protected-access
+            self.next_refresh_time = datetime.now() + timedelta(seconds=self.refresh_options.refresh_interval)
+            self.attempts = 1
+            self.refresh_options._callback()
+
+        def failed_update(self, error, message):
+            logging.warning(message, self.attempts, error)
+            # Refresh All or None, any failure will trigger a backoff
+            # pylint:disable=protected-access
+            self.next_refresh_time = datetime.now() + timedelta(microseconds=self.calculate_backoff())
+            self.attempts += 1
+            self.refresh_options._on_error()
+
+        def calculate_backoff(self):
+            max_attempts = 30
```

Comment:
why 30?

Reply:
This is the same value we have in the other libraries so we don't overflow the max number.

Reply:
You can make more attempts, it just won't increate the time anymore.

Reply:
It should be calculated based on how many times we can left shift 1 without overflowing.
Its 63 in .NET because we left shift `(long) 1` to get powers of 2. So its calculated as `(sizeof(long) * 8) - 1`
Can we calculate the same for Python here?